examples/ecs/state.rs

```diff
@@ -16,40 +15,84 @@ fn main() {
         // All systems from the exit schedule of the state we're leaving are run first,
         // and then all systems from the enter schedule of the state we're entering are run second.
         .add_systems(OnEnter(AppState::Menu), setup_menu)
+        // We also have the ability to use matchers, which are structs that allow more complex matching functions
+        // These can be created with either the `::matching` function, which will run if only one of the systems matches
+        // or with the `::matching_strict` function, which will run if the entering/exiting system matches
+        // regardless of whether the other system matches as well.
+        .add_systems(OnEnter::matching(InGame), setup_game)
+        .add_systems(OnExit::matching_strict(ShowsUI), cleanup_ui)
+        // You can also use the `on_enter`, `on_enter_strict`, `on_exit` or `on_exit_strict` macros
+        // The macros can either accept a matcher (or state) as a single parameter,
+        // or the state type and a pattern to match against as two parameters
+        .add_systems(on_enter!(AppState::InGame { paused: true }), setup_paused)
+        .add_systems(
+            on_enter!(AppState::InGame { paused: false }),
+            setup_in_game_ui,
+        )
         // By contrast, update systems are stored in the `Update` schedule. They simply
         // check the value of the `State<T>` resource to see if they should run each frame.
+        // The in_state function can accept State values or StateMatcher's.
         .add_systems(Update, menu.run_if(in_state(AppState::Menu)))
-        .add_systems(OnExit(AppState::Menu), cleanup_menu)
-        .add_systems(OnEnter(AppState::InGame), setup_game)
+        // Alternatively, you can use the `in_state!` macro with a type and pattern
+        .add_systems(
+            Update,
+            change_color.run_if(in_state!(AppState, AppState::InGame { .. })),
```

Comment:
Mentioned elsewhere, but unfortunately not... We can get the type of an expression, but not of a pattern. So you could do `in_state!(AppState::Menu)` since that is an expression, but for pattern matching it won't work.